pkg/operator/controllers/checkers/serviceprincipalchecker/checker.go

```diff
@@ -0,0 +1,61 @@
+package serviceprincipalchecker
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Azure/ARO-RP/pkg/api/validate/dynamic"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/clusterauthorizer"
+)
+
+type servicePrincipalChecker interface {
+	Check(ctx context.Context, AZEnvironment string) error
+}
+
+type checker struct {
+	log *logrus.Entry
+
+	credentialsGetter      func(ctx context.Context) (*clusterauthorizer.Credentials, error)
+	spValidatorConstructor func(azEnv *azureclient.AROEnvironment) (dynamic.ServicePrincipalValidator, error)
```

Comment:
are these required to be functions or could we use interfaces here instead? 
also, we probably could just have `spValidator` and delegate the construction to the implementer

Reply:
the `dynamic` package does weird things actually, we might want to look into it to fix it because the current implementation is unnecessarily complicated. e.g. this (which really seems to be an unneeded constructor, which returns the interface type instead of the actual type): 
```go 
func NewServicePrincipalValidator(log *logrus.Entry, azEnv *azureclient.AROEnvironment, authorizerType AuthorizerType, tokenClient aad.TokenClient) (ServicePrincipalValidator, error) {
	return &dynamic{
		log:            log,
		authorizerType: authorizerType,
		azEnv:          azEnv,
		tokenClient:    tokenClient,
	}, nil
}
```

but this is for another PR

Reply:
> are these required to be functions or could we use interfaces here instead?

These are needed so that we can swap constructors (`clusterauthorizer.AzCredentials(...)` and `dynamic.NewServicePrincipalValidator(...)`) with constructors which return mocks.

We can define an interface, but defining an interface for a constructure seems like an overkill to me.

> the `dynamic` package does weird things actually, we might want to look into it to fix it because the current implementation is unnecessarily complicated.

I agere with this part

> e.g. this (which really seems to be an unneeded constructor, which returns the interface type instead of the actual type)

I think nothing wrong with it: we return an interface becuase we don't want to expose fields of `dynamic`. If you need a validator somewhere - you use interface and mock the whole thing if you want to test your unit.

Reply:
but since it is for a another PR, that is why I was recommending to delegate it to an interface so we didn't have to/would be easier change it later when we fix `dynamic`

Reply:
> and this is why we end up having to swap out useless constructors.

No, it is not the reason. We have to do that because of two reasons:
* For the validator - info about the environment (e.g. public cloud, gov cloud) comes from an instance which we know only on the reconcile stage. If we can address this - we can drop a swappable constructor for the validator. For rest of the codebase - I think it can already be stored in the field. E.g. [here](https://github.com/Azure/ARO-RP/blob/d5485932f428cf2c61d51820a825b1c04fe97d20/pkg/api/validate/openshiftcluster_validatedynamic.go#L59-L63) we should be able to store it in a field of `openShiftClusterDynamicValidator`.
* For `credentialsGetter` - we can not store credentials between the reconcile calls becasue we want to ensure that whatever we have in the cluster is valid hence we need to call `clusterauthorizer.AzCredentials(...)` on every reconciliation. `credentialsGetter` simply makes it swappable so that we can mock the whole thing (`clusterauthorizer.AzCredentials` itself is tested separately).

Can we do it differently? Yes. (e.g. some sort of `Setup` method as you suggested). Does it make sense? I don't know... Probably not.

> Here if the servicePrincipal had a `Setup` function instead of a constructor (because the fields are private), we could just use the interface and avoid having this extra function field. and then we could just use a common mock for all the places where we use it to test it

Do you mean that you would prefer having validator as a field in the `checker` and do a `Setup` call on every reconcile? Yes, that would work, but you don't have to expose a sepcific type for this: `Setup` can be part of your interface.

However I personally wouldn't support this appraoch since this introduces unnecessary state management inside of the validaotr.

In my opinion swappable constructor which returns a disposable validator instance is a better appraoch than having a validator hanging around for the lifetime of the controller and having to deal with credentials replacement. I think it is harder to modify code which deals with the state like this: easier to introduce a regression.

> also, you chose a naming that is for interfaces, functions names should be `getCredentials` and `constructSpValidator` or `newSpValidator`

I agree that `newSpValidator` is probably better. `credentialsGetter` should probably just become `credentials` (so that we follow [this advice](https://go.dev/doc/effective_go#Getters)).

I would like to update it in a follow up PR so we don't have to wait for E2E here just because of naming, if that's ok with you?

---

Off topic:

> why don't we want to expose those types? As long as the fields in it are private, there isn't really any need to hide those fields.

But why do you want to do that? You have a contract between packages (an interface). Why should you care about a specific type/implementation details?

>  I think this is just a bad design on the dynamic part

I don't think so. Maybe I miss something, but I don't believe that returning an interface is a bad design. Your package provides an interface and consumers (even withing the same codebase) shouldn't care about the implementation details. You might want to swap the implementation one day or add another one which meets the interface and you don't want to update consuming code because of that.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#interfaces this says to return concrete types instead of interfaces 
```The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.```

Reply:
when I was saying that we should have a `Setup` function on the struct, I was thinking of adding it to the `ServicePrincipalValidator` interface. But that would require to not have the `dynamic` struct implementing all the interfaces at once. (to be completely honest, I don't understand why it is done this way, it makes it very cumbersome)